chia/types/eligible_coin_spends.py

```diff
@@ -97,12 +222,138 @@ def get_deduplication_info(
+                    eligible_for_fast_forward=spend_data.eligible_for_fast_forward,
+                    additions=patched_additions,
+                    cost=spend_data.cost,
+                )
+                # Update the list of coins spends that will make the new fast
+                # forward spend bundle
+                new_coin_spends.append(new_coin_spend)
+                # We're done here, moving on
+                continue
+            # We processed this puzzle hash before, so build on that
+            # See first if we're the most recent version
+            if unspent_lineage_info.coin_id == coin_id:
+                # We are, so we don't need to fast forward, we just need to
+                # find the next version among our additions to chain ff spends
+                set_next_singleton_version(
+                    current_singleton=spend_data.coin_spend.coin,
+                    singleton_additions=spend_data.additions,
+                    fast_forward_spends=self.fast_forward_spends,
+                )
+                # Nothing more to do for this spend, moving on
+                new_coin_spends.append(spend_data.coin_spend)
+                continue
+            # We're not the most recent version, so let's fast forward
+            new_coin_spend, patched_additions = perform_the_fast_forward(
+                unspent_lineage_info=unspent_lineage_info,
+                spend_data=spend_data,
+                fast_forward_spends=self.fast_forward_spends,
+            )
+            # Mark this coin for a coin spend data update
+            ff_bundle_coin_spends[coin_id] = BundleCoinSpend(
+                coin_spend=new_coin_spend,
+                eligible_for_dedup=spend_data.eligible_for_dedup,
+                eligible_for_fast_forward=spend_data.eligible_for_fast_forward,
+                additions=patched_additions,
+                cost=spend_data.cost,
+            )
+            # Update the list of coins spends that make the new fast forward bundle
+            new_coin_spends.append(new_coin_spend)
+        if len(ff_bundle_coin_spends) == 0:
+            # This item doesn't have any fast forward coins, nothing to do here
+            return
+        # Update the mempool item after validating the new spend bundle
+        new_sb = SpendBundle(
+            coin_spends=new_coin_spends, aggregated_signature=mempool_item.spend_bundle.aggregated_signature
+        )
+        # We need to run the new spend bundle to make sure it remains valid
+        generator = simple_solution_generator(new_sb)
+        new_npc_result = get_name_puzzle_conditions(
+            generator=generator,
+            max_cost=mempool_item.npc_result.cost,
+            mempool_mode=True,
+            height=height,
+            constants=constants,
+        )
+        if new_npc_result.error is not None:
+            raise ValueError("Mempool item became invalid after singleton fast forward.")
```

Comment:
does any exception here lead to eviction from the mempool?

Reply:
Yes. That's visible upstream from here (`create_bundle_from_mempool_items`).